rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,115 @@
+---
+feature: system-parameters
+authors:
+  - "Eric Fu"
+start_date: "2023/01/11"
+---
+
+# RFC: System Parameters on Meta
+
+## Motivation
+
+Currently, we have 2 kinds of configurations:
+
+- Command-line arguments
+- Configuration file `risingwave.toml`
+
+[Distinguish system config variables and cli options #5676](https://github.com/risingwavelabs/risingwave/issues/5676) clarifies the difference between CLI args and config file:
+
+> Intuitively, *Options* should contain basic user-facing options, e.g., addr & host and maybe some feature gates, while *Config* contains deep-dark internal system variables to fine-tune the system performance.
+
+However, this is not a complete solution because
+
+1. **`risingwave.toml` is a local file**, which makes it possible that multiple nodes could see different parameters and this is absolutely unacceptable. For example, if some nodes use `data_directory = 'hummock_001'` and others use `data_directory = 'hummock_002'`, the stored data will be corrupted.
+2. **Unable to `SET` a config after the process started**. Even if changes are written to disk, new nodes still miss these changes.
+
+This proposal will continue to improve the configuration mechanism but also repudiate the design of #5676.
+
+Previously I think this might not be an urgent issue until I realized that our Cloud platform couples tightly with more and more configurations. The later we commence, the heavier burden it will become. 
+
+## Design
+
+```mermaid
+graph LR
+    all[Does it apply to the RisingWave instance?]
+    all ==No==> process["Process Configs (Local)"]
+    all ==Yes==> system["System Configs (on Meta)"]
+    process --> cli_args(Command-line arguments)
+    process --> conf_file("Config file (.toml)")
+    system --> sql_cmd(SQL interface: SET / SHOW)
+    system --> cloud(Web Console)
+```
+
+### System Configs
+
+What we need is a centralized source-of-truth of cluster configurations. Obviously, the best choice for RisingWave is the Meta Service and the underlying etcd. 
+
+Here we could borrow the syntax from Postgres:
+
+```sql
+-- Set or reset a parameter
+ALTER SYSTEM / DATABASE SET parameter TO value / 'value' / DEFAULT;
```

Comment:
Another thing not discussed is session variables. `SET [ SESSION | LOCAL ] configuration_parameter` affects local setting, while `ALTER SYSTEM SET` affects global setting. From current design I think we simply don't allow session level config? 🤔 Or in other words, will we distinguish session variables from system variables? In PG it seems they are also the same set of variables.

I also found interesting notes in tidb doc https://docs.pingcap.com/tidb/stable/system-variables

> Several GLOBAL variables persist to the TiDB cluster. Some variables in this document have a Persists to cluster setting, which can be configured to Yes or No.

Reply:
> I found PG distinguishes SESSION variables and GLOBAL parameters clearly, so I decided not to discuss session variables as well.

Really? I thought PG distinguishes global parameters and session variables strictly.

From my understanding, they are not related for most cases, except one: the GLOBAL parameter decides the default value for new sessions, and a session can SET its parameter to override that default value.

Reply:
> Really? I thought PG distinguishes global parameters and session variables strictly, unlike MySQL/TiDB.

According to [20.1.3. Parameter Interaction via SQL](https://www.postgresql.org/docs/current/config-setting.html#CONFIG-SETTING-SQL-COMMAND-INTERACTION): system level, db level, session level configs seem to be a single set of variables 😇

Reply:
Yeah you are right ❤️. So I should discuss the session parameter as well.